File: instant-quote/src/lib/mesh/analyze.ts
// Geometry analysis from a position soup (9 floats/triangle, mm units).
// Single pass for signed volume, surface area, bbox, triangle count.
// Falls back to convex-hull volume when the mesh isn't watertight.

import type { MeshMetrics } from './types'
import { convexHullVolume, pointsFromPositions } from './hull'
import { MeshParseError } from './parse-stl'

export function analyze(positions: Float32Array): MeshMetrics {
  const triangleCount = Math.floor(positions.length / 9)
  if (triangleCount === 0) {
    throw new MeshParseError('empty', 'Mesh has no triangles.')
  }

  let signedVol6 = 0 // 6× signed volume, mm³
  let area2 = 0 // 2× area, mm²
  let minX = Infinity
  let minY = Infinity
  let minZ = Infinity
  let maxX = -Infinity
  let maxY = -Infinity
  let maxZ = -Infinity

  for (let i = 0; i < triangleCount * 9; i += 9) {
    const ax = positions[i]
    const ay = positions[i + 1]
    const az = positions[i + 2]
    const bx = positions[i + 3]
    const by = positions[i + 4]
    const bz = positions[i + 5]
    const cx = positions[i + 6]
    const cy = positions[i + 7]
    const cz = positions[i + 8]

    // Signed volume: dot(a, cross(b, c))
    const crx = by * cz - bz * cy
    const cry = bz * cx - bx * cz
    const crz = bx * cy - by * cx
    signedVol6 += ax * crx + ay * cry + az * crz

    // Area: |cross(b-a, c-a)| / 2
    const e1x = bx - ax
    const e1y = by - ay
    const e1z = bz - az
    const e2x = cx - ax
    const e2y = cy - ay
    const e2z = cz - az
    const nx = e1y * e2z - e1z * e2y
    const ny = e1z * e2x - e1x * e2z
    const nz = e1x * e2y - e1y * e2x
    area2 += Math.sqrt(nx * nx + ny * ny + nz * nz)

    if (ax < minX) minX = ax
    if (ay < minY) minY = ay
    if (az < minZ) minZ = az
    if (ax > maxX) maxX = ax
    if (ay > maxY) maxY = ay
    if (az > maxZ) maxZ = az
    if (bx < minX) minX = bx
    if (by < minY) minY = by
    if (bz < minZ) minZ = bz
    if (bx > maxX) maxX = bx
    if (by > maxY) maxY = by
    if (bz > maxZ) maxZ = bz
    if (cx < minX) minX = cx
    if (cy < minY) minY = cy
    if (cz < minZ) minZ = cz
    if (cx > maxX) maxX = cx
    if (cy > maxY) maxY = cy
    if (cz > maxZ) maxZ = cz
  }

  const bboxMm = {
    x: maxX - minX,
    y: maxY - minY,
    z: maxZ - minZ,
  }
  const bboxVolumeCm3 = (bboxMm.x * bboxMm.y * bboxMm.z) / 1000

  const signedVolumeCm3 = signedVol6 / 6 / 1000
  const rawSignedVolumeCm3 = signedVolumeCm3
  const absVolumeCm3 = Math.abs(signedVolumeCm3)

  // Watertight test: positive signed volume within the bounding box, and — for
  // meshes small enough to check cheaply — every edge shared by exactly two
  // triangles (2-manifold, i.e. actually closed). Skip the edge check on very
  // large meshes for performance; fall back to the sign+bbox heuristic there.
  const signOk = signedVolumeCm3 > 0 && absVolumeCm3 <= bboxVolumeCm3 * 1.001
  const manifold =
    triangleCount >= 200_000 ? true : isEdgeManifold(positions, triangleCount)
  const watertight = signOk && manifold

  let volumeCm3 = absVolumeCm3
  let usedHullFallback = false
  if (!watertight) {
    const hullVolMm3 = convexHullVolume(pointsFromPositions(positions))
    if (hullVolMm3 > 0) {
      volumeCm3 = hullVolMm3 / 1000
      usedHullFallback = true
    }
  }

  return {
    volumeCm3,
    rawSignedVolumeCm3,
    surfaceAreaCm2: area2 / 2 / 100,
    bboxMm,
    triangleCount,
    watertight,
    usedHullFallback,
  }
}

/** True when every edge is shared by exactly two triangles (closed 2-manifold). */
function isEdgeManifold(
  positions: Float32Array,
  triangleCount: number,
): boolean {
  const vertexId = new Map<string, number>()
  const idOf = (i: number): number => {
    const key = `${Math.round(positions[i] * 1e4)},${Math.round(positions[i + 1] * 1e4)},${Math.round(positions[i + 2] * 1e4)}`
    let id = vertexId.get(key)
    if (id === undefined) {
      id = vertexId.size
      vertexId.set(key, id)
    }
    return id
  }

  const edgeCount = new Map<string, number>()
  const bump = (u: number, v: number) => {
    const key = u < v ? `${u}_${v}` : `${v}_${u}`
    edgeCount.set(key, (edgeCount.get(key) ?? 0) + 1)
  }

  for (let t = 0; t < triangleCount; t++) {
    const i = t * 9
    const a = idOf(i)
    const b = idOf(i + 3)
    const c = idOf(i + 6)
    bump(a, b)
    bump(b, c)
    bump(c, a)
  }

  for (const count of edgeCount.values()) {
    if (count !== 2) return false
  }
  return true
}

File: backend/internal/mesh/analyze.go
// Geometry analysis from a position soup (9 floats/triangle, mm units).
// Single pass for signed volume, surface area, bbox, triangle count.
// Port of instant-quote/src/lib/mesh/analyze.ts — WITHOUT the convex-hull
// fallback: when a mesh is not watertight the TS client substitutes its hull
// volume, but here Watertight=false is reported and the caller decides.

package mesh

import "math"

const inf = math.MaxFloat64

func analyzePositions(positions []float32) (Metrics, error) {
	triangleCount := len(positions) / 9
	if triangleCount == 0 {
		return Metrics{}, &ParseError{ErrEmpty, "Mesh has no triangles."}
	}

	var signedVol6 float64 // 6x signed volume, mm^3
	var area2 float64      // 2x area, mm^2
	minX, minY, minZ := inf, inf, inf
	maxX, maxY, maxZ := -inf, -inf, -inf

	for i := 0; i < triangleCount*9; i += 9 {
		ax := float64(positions[i])
		ay := float64(positions[i+1])
		az := float64(positions[i+2])
		bx := float64(positions[i+3])
		by := float64(positions[i+4])
		bz := float64(positions[i+5])
		cx := float64(positions[i+6])
		cy := float64(positions[i+7])
		cz := float64(positions[i+8])

		// Signed volume: dot(a, cross(b, c))
		crx := by*cz - bz*cy
		cry := bz*cx - bx*cz
		crz := bx*cy - by*cx
		signedVol6 += ax*crx + ay*cry + az*crz

		// Area: |cross(b-a, c-a)| / 2
		e1x := bx - ax
		e1y := by - ay
		e1z := bz - az
		e2x := cx - ax
		e2y := cy - ay
		e2z := cz - az
		nx := e1y*e2z - e1z*e2y
		ny := e1z*e2x - e1x*e2z
		nz := e1x*e2y - e1y*e2x
		area2 += math.Sqrt(nx*nx + ny*ny + nz*nz)

		for _, p := range [3][3]float64{{ax, ay, az}, {bx, by, bz}, {cx, cy, cz}} {
			if p[0] < minX {
				minX = p[0]
			}
			if p[0] > maxX {
				maxX = p[0]
			}
			if p[1] < minY {
				minY = p[1]
			}
			if p[1] > maxY {
				maxY = p[1]
			}
			if p[2] < minZ {
				minZ = p[2]
			}
			if p[2] > maxZ {
				maxZ = p[2]
			}
		}
	}

	bboxMm := Vec3{X: maxX - minX, Y: maxY - minY, Z: maxZ - minZ}
	bboxVolumeCm3 := (bboxMm.X * bboxMm.Y * bboxMm.Z) / 1000

	signedVolumeCm3 := signedVol6 / 6 / 1000
	absVolumeCm3 := math.Abs(signedVolumeCm3)

	// Watertight test: positive signed volume within the bounding box, and —
	// for meshes small enough to check cheaply — every edge shared by exactly
	// two triangles (2-manifold, i.e. actually closed). Skip the edge check on
	// very large meshes for performance.
	signOk := signedVolumeCm3 > 0 && absVolumeCm3 <= bboxVolumeCm3*1.001
	manifold := triangleCount >= 200_000 || isEdgeManifold(positions, triangleCount)
	watertight := signOk && manifold

	return Metrics{
		VolumeCm3:          absVolumeCm3,
		RawSignedVolumeCm3: signedVolumeCm3,
		SurfaceAreaCm2:     area2 / 2 / 100,
		BboxMm:             bboxMm,
		TriangleCount:      triangleCount,
		Watertight:         watertight,
		UsedHullFallback:   false, // Go has no hull fallback
	}, nil
}

// isEdgeManifold reports whether every edge is shared by exactly two triangles
// (closed 2-manifold). Vertices are quantized to 1e-4 mm before comparison.
func isEdgeManifold(positions []float32, triangleCount int) bool {
	type key [3]int64
	vertexID := map[key]int{}
	idOf := func(i int) int {
		k := key{
			int64(math.Round(float64(positions[i]) * 1e4)),
			int64(math.Round(float64(positions[i+1]) * 1e4)),
			int64(math.Round(float64(positions[i+2]) * 1e4)),
		}
		id, ok := vertexID[k]
		if !ok {
			id = len(vertexID)
			vertexID[k] = id
		}
		return id
	}

	edgeCount := map[[2]int]int{}
	bump := func(u, v int) {
		if u > v {
			u, v = v, u
		}
		edgeCount[[2]int{u, v}]++
	}

	for t := 0; t < triangleCount; t++ {
		i := t * 9
		a := idOf(i)
		b := idOf(i + 3)
		c := idOf(i + 6)
		bump(a, b)
		bump(b, c)
		bump(c, a)
	}

	for _, count := range edgeCount {
		if count != 2 {
			return false
		}
	}
	return true
}
